File: exam_prep/IST603/notebook_app/notebook.py
"""
Notebook class per UML: manages a list of Note; search, new_note, modify_memo, modify_tags.
"""

from typing import List

from note import Note


class Notebook:
    """Manages a collection of Note objects."""

    def __init__(self) -> None:
        self.notes: List[Note] = []
        self._next_id = 1

    def search(self, filter_str: str) -> List[Note]:
        """Return a list of notes that match the filter."""
        return [n for n in self.notes if n.match(filter_str)]

    def new_note(self, memo: str, tags: str = "") -> Note:
        """Create a new note and add it to the notebook."""
        note = Note(self._next_id, memo, tags)
        self._next_id += 1
        self.notes.append(note)
        return note

    def modify_memo(self, note_id: int, memo: str) -> None:
        """Update the memo of the note with the given id."""
        note = self._find_by_id(note_id)
        if note:
            note.memo = memo

    def modify_tags(self, note_id: int, tags: str) -> None:
        """Update the tags of the note with the given id."""
        note = self._find_by_id(note_id)
        if note:
            note.tags = tags

    def _find_by_id(self, note_id: int):
        """Return the Note with the given id, or None."""
        for n in self.notes:
            if n.note_id == note_id:
                return n
        return None

File: exam_prep/IST603/notebook_app/java/ist603/notebook/Notebook.java
package ist603.notebook;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Notebook per UML: manages a list of Note; search, new_note, modify_memo, modify_tags.
 */
public class Notebook {
    public List<Note> notes = new ArrayList<>();
    private int nextId = 1;

    /**
     * Returns a list of notes that match the filter.
     */
    public List<Note> search(String filter) {
        if (filter == null) {
            return new ArrayList<>(notes);
        }
        return notes.stream()
                .filter(n -> n.match(filter))
                .collect(Collectors.toList());
    }

    /**
     * Creates a new note and adds it to the notebook.
     */
    public Note newNote(String memo, String tags) {
        String t = tags != null ? tags : "";
        Note note = new Note(nextId++, memo, t);
        notes.add(note);
        return note;
    }

    /**
     * Overload with default empty tags.
     */
    public Note newNote(String memo) {
        return newNote(memo, "");
    }

    /**
     * Updates the memo of the note with the given id.
     */
    public void modifyMemo(int noteId, String memo) {
        Note note = findById(noteId);
        if (note != null) {
            note.memo = memo;
        }
    }

    /**
     * Updates the tags of the note with the given id.
     */
    public void modifyTags(int noteId, String tags) {
        Note note = findById(noteId);
        if (note != null) {
            note.tags = tags != null ? tags : "";
        }
    }

    private Note findById(int noteId) {
        for (Note n : notes) {
            if (n.getNoteId() == noteId) {
                return n;
            }
        }
        return null;
    }
}
